File: iOS/G3MiOSSDK/Commons/Geometry/Geodetic2D.cpp
//
//  Geodetic3D.cpp
//  G3MiOSSDK
//
//  Created by Diego Gomez Deck on 31/05/12.
//

#include "Geodetic2D.hpp"
#include "Geodetic3D.hpp"

#include "IStringBuilder.hpp"
#include "Vector3D.hpp"
#include "IMathUtils.hpp"

Geodetic2D Geodetic2D::Geo2DNull = Geodetic2D::fromDegrees(NAND, NAND);

bool Geodetic2D::isBetween(const Geodetic2D& min,
                           const Geodetic2D& max) const {
  return
  _latitude.isBetween(min._latitude, max._latitude) &&
  _longitude.isBetween(min._longitude, max._longitude);
}

bool Geodetic2D::closeTo(const Geodetic2D &other) const {
  if (!_latitude.closeTo(other._latitude)) {
    return false;
  }

  return _longitude.closeTo(other._longitude);
}

const std::string Geodetic2D::description() const {
  IStringBuilder* isb = IStringBuilder::newStringBuilder();
  isb->addString("(lat=");
  isb->addString(_latitude.description());
  isb->addString(", lon=");
  isb->addString(_longitude.description());
  isb->addString(")");
  const std::string s = isb->getString();
  delete isb;
  return s;
}

Angle Geodetic2D::angleTo(const Geodetic2D& that) const {
  const double cos1 = COS(_latitude._radians);
  const Vector3D normal1(cos1 * COS(_longitude._radians),
                         cos1 * SIN(_longitude._radians),
                         SIN(_latitude._radians));
  const double cos2 = COS(that._latitude._radians);
  const Vector3D normal2(cos2 * COS(that._longitude._radians),
                         cos2 * SIN(that._longitude._radians),
                         SIN(that._latitude._radians));

  return Angle::fromRadians(asin(normal1.cross(normal2).squaredLength()));
}

double Geodetic2D::bearingInRadians(const Angle& fromLatitude,
                                    const Angle& fromLongitude,
                                    const Angle& toLatitude,
                                    const Angle& toLongitude) {
  const IMathUtils* mu = IMathUtils::instance();

  const double deltaLonRad = toLongitude._radians - fromLongitude._radians;

  const double toLatCos = COS(toLatitude._radians);

  const double y = SIN(deltaLonRad) * toLatCos;
  const double x = COS(fromLatitude._radians) * SIN(toLatitude._radians) - SIN(fromLatitude._radians) * toLatCos * COS(deltaLonRad);
  const double radians = mu->atan2(y, x);
  return radians;

  //  const double pi2 = PI*2;
  //  return mu->mod(radians + pi2, pi2);

  //  const double r1 = mu->mod(radians, pi2);
  //  const double r2 = mu->mod(radians + pi2, pi2);
  //  return (mu->abs(r1) < mu->abs(r2)) ? r1 : r2;
}

double Geodetic2D::bearingInDegrees(const Angle& fromLatitude,
                                    const Angle& fromLongitude,
                                    const Angle& toLatitude,
                                    const Angle& toLongitude) {
  return TO_DEGREES(bearingInRadians(fromLatitude,
                                     fromLongitude,
                                     toLatitude,
                                     toLongitude));
}

Geodetic3D Geodetic2D::asGeodetic3DWithHeight(double height) const{
  return Geodetic3D(_latitude, _longitude, height);
}

File: tools/vectorial-streaming/src/com/glob3mobile/geo/Geodetic2D.java


package com.glob3mobile.geo;

import java.io.Serializable;


public class Geodetic2D
   implements
      Serializable {


   private static final long serialVersionUID = 1L;


   public static Geodetic2D fromRadians(final double lat,
                                        final double lon) {
      return new Geodetic2D(Angle.fromRadians(lat), Angle.fromRadians(lon));
   }


   public static Geodetic2D fromDegrees(final double lat,
                                        final double lon) {
      return new Geodetic2D(Angle.fromDegrees(lat), Angle.fromDegrees(lon));
   }


   public final Angle _latitude;
   public final Angle _longitude;


   public Geodetic2D(final Angle latitude,
                     final Angle longitude) {
      _latitude = latitude;
      _longitude = longitude;
   }


   @Override
   public String toString() {
      return "[" + _latitude + ", " + _longitude + "]";
   }


   @Override
   public int hashCode() {
      final int prime = 31;
      int result = 1;
      result = (prime * result) + ((_latitude == null) ? 0 : _latitude.hashCode());
      result = (prime * result) + ((_longitude == null) ? 0 : _longitude.hashCode());
      return result;
   }


   @Override
   public boolean equals(final Object obj) {
      if (this == obj) {
         return true;
      }
      if (obj == null) {
         return false;
      }
      if (getClass() != obj.getClass()) {
         return false;
      }
      final Geodetic2D other = (Geodetic2D) obj;
      if (_latitude == null) {
         if (other._latitude != null) {
            return false;
         }
      }
      else if (!_latitude.equals(other._latitude)) {
         return false;
      }
      if (_longitude == null) {
         if (other._longitude != null) {
            return false;
         }
      }
      else if (!_longitude.equals(other._longitude)) {
         return false;
      }
      return true;
   }


   public boolean closeTo(final Geodetic2D that) {
      return _latitude.closeTo(that._latitude) && _longitude.closeTo(that._longitude);
   }


   public String toName() {
      return _latitude.toName() + "_" + _longitude.toName();
   }


}
